modules/ml/src/main/java/org/apache/ignite/ml/dataset/impl/local/LocalDatasetBuilder.java

```diff
@@ -99,37 +122,28 @@ public LocalDatasetBuilder(Map<K, V> upstreamMap, IgniteBiPredicate<K, V> filter
 
         int ptr = 0;
 
+        List<LearningEnvironment> envs = IntStream.range(0, partitions).boxed().map(envBuilder::buildForWorker)
+            .collect(Collectors.toList());
+
         for (int part = 0; part < partitions; part++) {
             int cnt = part == partitions - 1 ? entriesList.size() - ptr : Math.min(partSize, entriesList.size() - ptr);
+            LearningEnvironment env = envs.get(part);
+            UpstreamTransformer<K, V> transformer1 = upstreamTransformerBuilder.build(env);
+            UpstreamTransformer<K, V> transformer2 = Utils.copy(transformer1);
```

Comment:
Why do we need a copy here?

Reply:
Because we can possibly have not a stateless transformer. For now, we have only stateless transformer (`BaggingUpstreamTransformer`), but maybe in the future... On the other hand, we can add some contracts in documentation banning states and get rid of this clumsy copying.